test/TestInternalGrains/StreamingImmutabilityTestGrain.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans;
+using Orleans.Streams;
+using UnitTests.GrainInterfaces;
+
+namespace UnitTests.Grains
+{
+    public class StreamingImmutabilityTestGrain : Grain, IStreamingImmutabilityTestGrain
+    {
+        private StreamImmutabilityTestObject _myObject;
+        private StreamSubscriptionHandle<StreamImmutabilityTestObject> _streamSubscriptionHandle;
+
+        public async Task SubscribeToStream(Guid guid, string providerName)
+        {
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(guid, "Namespace");
+            _streamSubscriptionHandle = await stream.SubscribeAsync(OnNextAsync);
+        }
+
+        public async Task UnsubscribeFromStream()
+        {
+            if (_streamSubscriptionHandle != null)
+                await _streamSubscriptionHandle.UnsubscribeAsync();
+        }
+
+        public async Task ProduceTestObject(string providerName)
+        {
+            _myObject = new StreamImmutabilityTestObject() {MyString = this.GetPrimaryKey().ToString()};
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(this.GetPrimaryKey(), "Namespace");
+            await stream.OnNextAsync(_myObject);
```

Comment:
> Or would you prefer to just test the immutability behavior with a local consumer and not consider the remote consumer case

I think this is the simplest test which gets at the heart of the behavior.  If we make an OnNextAsync call, then immediately change the object state before resolving the task, the delivered event (regardless of locality) should have the original value when immutability is off, and the modified value when immutability is on.